In [10]:
import pandas as pd

df = pd.read_csv("weather_data.csv")

In [11]:
#inspecting the structure and content of a DataFrame to understand the data it contains
df.head()

loc_id        date  time isdaytime  tempC  tempF  windspeedMiles  \
0       1  2019-06-01  12AM        no     16     61               6   
1       1  2019-06-01   1AM        no     16     60               6   
2       1  2019-06-01   2AM        no     15     59               5   
3       1  2019-06-01   3AM        no     15     58               5   
4       1  2019-06-01   4AM        no     15     59               4   

   windspeedKmph  winddirdegree winddir16point  ...  HeatIndexC HeatIndexF  \
0             10            228             SW  ...          16         61   
1              9            230             SW  ...          16         60   
2              9            233             SW  ...          15         59   
3              8            235             SW  ...          15         58   
4              7            229             SW  ...          15         59   

  DewPointC  DewPointF  WindChillC  WindChillF  WindGustMiles  WindGustKmph  \
0        12         54          16          61              7            12   
1        12         53          16          60              7            11   
2        11         53          15          59              6            10   
3        11         52          14          58              6             9   
4        11         52          15          59              5             8   

   FeelsLikeC  FeelsLikeF  
0          16          61  
1          16          60  
2          15          59  
3          14          58  
4          15          59  

[5 rows x 31 columns]

In [28]:
#examine column names
df.columns

Index(['loc_id', 'date', 'time', 'isdaytime', 'tempC', 'tempF',
       'windspeedMiles', 'windspeedKmph', 'winddirdegree', 'winddir16point',
       'weatherCode', 'weatherIconUrl', 'weatherDesc', 'precipMM',
       'precipInches', 'humidity', 'visibilityKm', 'visibilityMiles',
       'pressureMB', 'pressureInches', 'cloudcover', 'HeatIndexC',
       'HeatIndexF', 'DewPointC', 'DewPointF', 'WindChillC', 'WindChillF',
       'WindGustMiles', 'WindGustKmph', 'FeelsLikeC', 'FeelsLikeF'],
      dtype='object')

In [13]:
#check whether a specific column in a DataFrame contains numeric data
def is_numeric(col):
    try:
        pd.to_numeric(col)
        return True
    except:
        return False

In [14]:
#Find all the numerical columns
numeric_columns = []
for column in df.columns:
  if(is_numeric(df[column])):
    numeric_columns.append(column)

In [15]:
print(numeric_columns)

['loc_id', 'tempC', 'tempF', 'windspeedMiles', 'windspeedKmph', 'winddirdegree', 'weatherCode', 'precipMM', 'precipInches', 'humidity', 'visibilityKm', 'visibilityMiles', 'pressureMB', 'pressureInches', 'cloudcover', 'HeatIndexC', 'HeatIndexF', 'DewPointC', 'DewPointF', 'WindChillC', 'WindChillF', 'WindGustMiles', 'WindGustKmph', 'FeelsLikeC', 'FeelsLikeF']


In [16]:
#normalization of nuerical data i.e between 0 and 1
#MinMax Scaling for all umeric colums
#MinMAx scaling allows us to get information about the significance of each entry in a column
df_min_max_scaled = df.copy()
for numeric_column in numeric_columns:
  df_min_max_scaled[numeric_column] = (df_min_max_scaled[numeric_column] - df_min_max_scaled[numeric_column].min()) / (df_min_max_scaled[numeric_column].max() - df_min_max_scaled[numeric_column].min())

In [17]:
df_min_max_scaled.head()
df_min_max_scaled.drop(["loc_id"], axis=1)
# We donot need the location id at this point

date  time isdaytime     tempC     tempF  windspeedMiles  \
0     2019-06-01  12AM        no  0.366667  0.370370        0.285714   
1     2019-06-01   1AM        no  0.366667  0.351852        0.285714   
2     2019-06-01   2AM        no  0.333333  0.333333        0.238095   
3     2019-06-01   3AM        no  0.333333  0.314815        0.238095   
4     2019-06-01   4AM        no  0.333333  0.333333        0.190476   
...          ...   ...       ...       ...       ...             ...   
1459  2019-07-31   7PM        no  0.500000  0.500000        0.523810   
1460  2019-07-31   8PM        no  0.500000  0.481481        0.523810   
1461  2019-07-31   9PM        no  0.466667  0.462963        0.476190   
1462  2019-07-31  10PM        no  0.433333  0.444444        0.428571   
1463  2019-07-31  11PM        no  0.433333  0.425926        0.380952   

      windspeedKmph  winddirdegree winddir16point  weatherCode  ...  \
0           0.28125       0.640227             SW     0.010870  ...   
1           0.25000       0.645892             SW     0.010870  ...   
2           0.25000       0.654391             SW     0.010870  ...   
3           0.21875       0.660057             SW     0.010870  ...   
4           0.18750       0.643059             SW     0.010870  ...   
...             ...            ...            ...          ...  ...   
1459        0.53125       0.730878              W     0.228261  ...   
1460        0.50000       0.725212            WSW     0.010870  ...   
1461        0.43750       0.722380            WSW     0.010870  ...   
1462        0.43750       0.725212            WSW     0.010870  ...   
1463        0.40625       0.728045            WSW     0.010870  ...   

     HeatIndexC HeatIndexF  DewPointC  DewPointF  WindChillC  WindChillF  \
0      0.259259   0.255319     0.5000   0.500000    0.366667    0.370370   
1      0.259259   0.234043     0.5000   0.464286    0.366667    0.351852   
2      0.222222   0.212766     0.4375   0.464286    0.333333    0.333333   
3      0.222222   0.191489     0.4375   0.428571    0.300000    0.314815   
4      0.222222   0.212766     0.4375   0.428571    0.333333    0.333333   
...         ...        ...        ...        ...         ...         ...   
1459   0.407407   0.404255     0.2500   0.250000    0.500000    0.500000   
1460   0.407407   0.382979     0.3125   0.285714    0.500000    0.481481   
1461   0.370370   0.361702     0.3750   0.321429    0.466667    0.462963   
1462   0.333333   0.340426     0.3750   0.357143    0.433333    0.444444   
1463   0.333333   0.319149     0.4375   0.428571    0.433333    0.425926   

      WindGustMiles  WindGustKmph  FeelsLikeC  FeelsLikeF  
0           0.21875          0.22    0.354839    0.363636  
1           0.21875          0.20    0.354839    0.345455  
2           0.18750          0.18    0.322581    0.327273  
3           0.18750          0.16    0.290323    0.309091  
4           0.15625          0.14    0.322581    0.327273  
...             ...           ...         ...         ...  
1459        0.43750          0.44    0.483871    0.490909  
1460        0.43750          0.42    0.483871    0.472727  
1461        0.40625          0.42    0.451613    0.454545  
1462        0.37500          0.38    0.419355    0.436364  
1463        0.34375          0.34    0.419355    0.418182  

[1464 rows x 30 columns]

In [18]:
#Total number of unique values in a column
value_counts = []
value_counts_dict = {}
for column in numeric_columns:
  value_counts_dict[column] = len(set(df_min_max_scaled[column]))
  value_counts.append(len(set(df_min_max_scaled[column])))

print(value_counts_dict)

{'loc_id': 1464, 'tempC': 30, 'tempF': 48, 'windspeedMiles': 22, 'windspeedKmph': 33, 'winddirdegree': 326, 'weatherCode': 15, 'precipMM': 35, 'precipInches': 5, 'humidity': 66, 'visibilityKm': 19, 'visibilityMiles': 12, 'pressureMB': 34, 'pressureInches': 2, 'cloudcover': 101, 'HeatIndexC': 28, 'HeatIndexF': 45, 'DewPointC': 17, 'DewPointF': 29, 'WindChillC': 30, 'WindChillF': 51, 'WindGustMiles': 32, 'WindGustKmph': 47, 'FeelsLikeC': 31, 'FeelsLikeF': 50}


In [19]:
value_sigificance = []
for value in value_counts:
  value_sigificance.append(value/sum(value_counts))

In [20]:
#Minmax normalization of column unique value count
# this will enable us to get the significance of each individual column

import numpy as np

data = np.array(value_sigificance)
normalizedData = (data-np.min(data))/(np.max(data)-np.min(data))
print(normalizedData)

[1.         0.01915185 0.03146375 0.01367989 0.02120383 0.22161423
 0.00889193 0.02257182 0.00205198 0.04377565 0.01162791 0.00683995
 0.02188782 0.         0.06771546 0.01778386 0.02941176 0.01025992
 0.01846785 0.01915185 0.03351573 0.02051984 0.03077975 0.01983584
 0.03283174]


In [21]:
df_min_max_scaled

loc_id        date  time isdaytime     tempC     tempF  windspeedMiles  \
0        NaN  2019-06-01  12AM        no  0.366667  0.370370        0.285714   
1        NaN  2019-06-01   1AM        no  0.366667  0.351852        0.285714   
2        NaN  2019-06-01   2AM        no  0.333333  0.333333        0.238095   
3        NaN  2019-06-01   3AM        no  0.333333  0.314815        0.238095   
4        NaN  2019-06-01   4AM        no  0.333333  0.333333        0.190476   
...      ...         ...   ...       ...       ...       ...             ...   
1459     NaN  2019-07-31   7PM        no  0.500000  0.500000        0.523810   
1460     NaN  2019-07-31   8PM        no  0.500000  0.481481        0.523810   
1461     NaN  2019-07-31   9PM        no  0.466667  0.462963        0.476190   
1462     NaN  2019-07-31  10PM        no  0.433333  0.444444        0.428571   
1463     NaN  2019-07-31  11PM        no  0.433333  0.425926        0.380952   

      windspeedKmph  winddirdegree winddir16point  ...  HeatIndexC HeatIndexF  \
0           0.28125       0.640227             SW  ...    0.259259   0.255319   
1           0.25000       0.645892             SW  ...    0.259259   0.234043   
2           0.25000       0.654391             SW  ...    0.222222   0.212766   
3           0.21875       0.660057             SW  ...    0.222222   0.191489   
4           0.18750       0.643059             SW  ...    0.222222   0.212766   
...             ...            ...            ...  ...         ...        ...   
1459        0.53125       0.730878              W  ...    0.407407   0.404255   
1460        0.50000       0.725212            WSW  ...    0.407407   0.382979   
1461        0.43750       0.722380            WSW  ...    0.370370   0.361702   
1462        0.43750       0.725212            WSW  ...    0.333333   0.340426   
1463        0.40625       0.728045            WSW  ...    0.333333   0.319149   

     DewPointC  DewPointF  WindChillC  WindChillF  WindGustMiles  \
0       0.5000   0.500000    0.366667    0.370370        0.21875   
1       0.5000   0.464286    0.366667    0.351852        0.21875   
2       0.4375   0.464286    0.333333    0.333333        0.18750   
3       0.4375   0.428571    0.300000    0.314815        0.18750   
4       0.4375   0.428571    0.333333    0.333333        0.15625   
...        ...        ...         ...         ...            ...   
1459    0.2500   0.250000    0.500000    0.500000        0.43750   
1460    0.3125   0.285714    0.500000    0.481481        0.43750   
1461    0.3750   0.321429    0.466667    0.462963        0.40625   
1462    0.3750   0.357143    0.433333    0.444444        0.37500   
1463    0.4375   0.428571    0.433333    0.425926        0.34375   

      WindGustKmph  FeelsLikeC  FeelsLikeF  
0             0.22    0.354839    0.363636  
1             0.20    0.354839    0.345455  
2             0.18    0.322581    0.327273  
3             0.16    0.290323    0.309091  
4             0.14    0.322581    0.327273  
...            ...         ...         ...  
1459          0.44    0.483871    0.490909  
1460          0.42    0.483871    0.472727  
1461          0.42    0.451613    0.454545  
1462          0.38    0.419355    0.436364  
1463          0.34    0.419355    0.418182  

[1464 rows x 31 columns]

In [22]:
# intended to store the significance values of columns in the dataset
column_significance = {}

for i in range(len(normalizedData)):
  column_significance[numeric_columns[i]] = normalizedData[i]

column_significance

{'loc_id': 1.0,
 'tempC': 0.019151846785225718,
 'tempF': 0.03146374829001368,
 'windspeedMiles': 0.013679890560875513,
 'windspeedKmph': 0.02120383036935705,
 'winddirdegree': 0.2216142270861833,
 'weatherCode': 0.008891928864569083,
 'precipMM': 0.0225718194254446,
 'precipInches': 0.0020519835841313274,
 'humidity': 0.04377564979480164,
 'visibilityKm': 0.011627906976744186,
 'visibilityMiles': 0.006839945280437757,
 'pressureMB': 0.02188782489740082,
 'pressureInches': 0.0,
 'cloudcover': 0.0677154582763338,
 'HeatIndexC': 0.017783857729138167,
 'HeatIndexF': 0.029411764705882353,
 'DewPointC': 0.010259917920656635,
 'DewPointF': 0.018467852257181942,
 'WindChillC': 0.019151846785225718,
 'WindChillF': 0.033515731874145,
 'WindGustMiles': 0.02051983584131327,
 'WindGustKmph': 0.0307797537619699,
 'FeelsLikeC': 0.019835841313269494,
 'FeelsLikeF': 0.03283173734610123}

In [29]:
#calculate the significance of a specific entry within a column
def get_entry_significance(parent_col, entry):
  if(str(entry) == 'nan'):
    return 0
  all_entries = len(df_min_max_scaled[parent_col])
  items = df_min_max_scaled[parent_col].value_counts()[entry]
  return items/all_entries


In [24]:
#GET sigificace of each etry in the 'table'
#We repeat the same techniques we used for the columns
entry_significance = {}
entry_significance_list = []
for column in numeric_columns:
  items = df_min_max_scaled[column]
  respective = []
  for item in items:
    respective.append(get_entry_significance(column,item))
  entry_significance_list.append(respective)
  entry_significance[column] = respective

In [25]:
for column in numeric_columns:
  df_min_max_scaled[column] = entry_significance[column]

In [33]:
df_min_max_scaled.head()
#Unscaled significance values

loc_id        date  time isdaytime     tempC     tempF  windspeedMiles  \
0     NaN  2019-06-01  12AM        no  0.659218  0.768421        0.829787   
1     NaN  2019-06-01   1AM        no  0.659218  0.536842        0.829787   
2     NaN  2019-06-01   2AM        no  0.469274  0.515789        0.675532   
3     NaN  2019-06-01   3AM        no  0.469274  0.484211        0.675532   
4     NaN  2019-06-01   4AM        no  0.469274  0.515789        0.579787   

   windspeedKmph  winddirdegree winddir16point  ...  HeatIndexC HeatIndexF  \
0       0.596026       0.652174             SW  ...    0.643243   0.760417   
1       0.609272       0.347826             SW  ...    0.643243   0.500000   
2       0.609272       0.391304             SW  ...    0.432432   0.510417   
3       0.503311       0.434783             SW  ...    0.432432   0.458333   
4       0.397351       0.565217             SW  ...    0.432432   0.510417   

  DewPointC  DewPointF  WindChillC  WindChillF  WindGustMiles  WindGustKmph  \
0  0.880829      0.675    0.614525    0.736842       0.593407      0.686957   
1  0.880829      0.900    0.614525    0.452632       0.593407      0.530435   
2  0.937824      0.900    0.351955    0.400000       0.637363      0.634783   
3  0.937824      0.825    0.329609    0.326316       0.637363      0.495652   
4  0.937824      0.825    0.351955    0.400000       0.412088      0.469565   

   FeelsLikeC  FeelsLikeF  
0    0.614525    0.736842  
1    0.614525    0.452632  
2    0.351955    0.400000  
3    0.329609    0.326316  
4    0.351955    0.400000  

[5 rows x 31 columns]

In [34]:
#Normalize values to be between 0 and 1
for numeric_column in numeric_columns:
  df_min_max_scaled[numeric_column] = (df_min_max_scaled[numeric_column] - df_min_max_scaled[numeric_column].min()) / (df_min_max_scaled[numeric_column].max() - df_min_max_scaled[numeric_column].min())

In [35]:
df_min_max_scaled.drop(["loc_id"], axis=1)

date  time isdaytime     tempC     tempF  windspeedMiles  \
0     2019-06-01  12AM        no  0.659218  0.768421        0.829787   
1     2019-06-01   1AM        no  0.659218  0.536842        0.829787   
2     2019-06-01   2AM        no  0.469274  0.515789        0.675532   
3     2019-06-01   3AM        no  0.469274  0.484211        0.675532   
4     2019-06-01   4AM        no  0.469274  0.515789        0.579787   
...          ...   ...       ...       ...       ...             ...   
1459  2019-07-31   7PM        no  0.731844  0.705263        0.563830   
1460  2019-07-31   8PM        no  0.731844  1.000000        0.563830   
1461  2019-07-31   9PM        no  0.916201  0.926316        0.734043   
1462  2019-07-31  10PM        no  0.804469  0.947368        0.734043   
1463  2019-07-31  11PM        no  0.804469  0.821053        0.872340   

      windspeedKmph  winddirdegree winddir16point  weatherCode  ...  \
0          0.596026       0.652174             SW     1.000000  ...   
1          0.609272       0.347826             SW     1.000000  ...   
2          0.609272       0.391304             SW     1.000000  ...   
3          0.503311       0.434783             SW     1.000000  ...   
4          0.397351       0.565217             SW     1.000000  ...   
...             ...            ...            ...          ...  ...   
1459       0.390728       0.130435              W     0.238683  ...   
1460       0.523179       0.478261            WSW     1.000000  ...   
1461       0.536424       0.217391            WSW     1.000000  ...   
1462       0.536424       0.478261            WSW     1.000000  ...   
1463       0.556291       0.130435            WSW     1.000000  ...   

     HeatIndexC HeatIndexF  DewPointC  DewPointF  WindChillC  WindChillF  \
0      0.643243   0.760417   0.880829   0.675000    0.614525    0.736842   
1      0.643243   0.500000   0.880829   0.900000    0.614525    0.452632   
2      0.432432   0.510417   0.937824   0.900000    0.351955    0.400000   
3      0.432432   0.458333   0.937824   0.825000    0.329609    0.326316   
4      0.432432   0.510417   0.937824   0.825000    0.351955    0.400000   
...         ...        ...        ...        ...         ...         ...   
1459   0.556757   0.520833   0.725389   0.733333    0.731844    0.705263   
1460   0.556757   0.937500   0.725389   0.525000    0.731844    1.000000   
1461   0.864865   0.906250   1.000000   0.916667    0.916201    0.926316   
1462   0.767568   0.916667   1.000000   0.916667    0.798883    0.947368   
1463   0.767568   0.812500   0.937824   0.825000    0.798883    0.810526   

      WindGustMiles  WindGustKmph  FeelsLikeC  FeelsLikeF  
0          0.593407      0.686957    0.614525    0.736842  
1          0.593407      0.530435    0.614525    0.452632  
2          0.637363      0.634783    0.351955    0.400000  
3          0.637363      0.495652    0.329609    0.326316  
4          0.412088      0.469565    0.351955    0.400000  
...             ...           ...         ...         ...  
1459       0.362637      0.365217    0.731844    0.705263  
1460       0.362637      0.478261    0.731844    1.000000  
1461       0.510989      0.478261    0.916201    0.926316  
1462       0.626374      0.443478    0.798883    0.947368  
1463       0.532967      0.513043    0.798883    0.810526  

[1464 rows x 30 columns]

In [36]:
#now we have a table populated with significance values that have een scaled to be between 0 and 1
df_min_max_scaled.head()

loc_id        date  time isdaytime     tempC     tempF  windspeedMiles  \
0     NaN  2019-06-01  12AM        no  0.659218  0.768421        0.829787   
1     NaN  2019-06-01   1AM        no  0.659218  0.536842        0.829787   
2     NaN  2019-06-01   2AM        no  0.469274  0.515789        0.675532   
3     NaN  2019-06-01   3AM        no  0.469274  0.484211        0.675532   
4     NaN  2019-06-01   4AM        no  0.469274  0.515789        0.579787   

   windspeedKmph  winddirdegree winddir16point  ...  HeatIndexC HeatIndexF  \
0       0.596026       0.652174             SW  ...    0.643243   0.760417   
1       0.609272       0.347826             SW  ...    0.643243   0.500000   
2       0.609272       0.391304             SW  ...    0.432432   0.510417   
3       0.503311       0.434783             SW  ...    0.432432   0.458333   
4       0.397351       0.565217             SW  ...    0.432432   0.510417   

  DewPointC  DewPointF  WindChillC  WindChillF  WindGustMiles  WindGustKmph  \
0  0.880829      0.675    0.614525    0.736842       0.593407      0.686957   
1  0.880829      0.900    0.614525    0.452632       0.593407      0.530435   
2  0.937824      0.900    0.351955    0.400000       0.637363      0.634783   
3  0.937824      0.825    0.329609    0.326316       0.637363      0.495652   
4  0.937824      0.825    0.351955    0.400000       0.412088      0.469565   

   FeelsLikeC  FeelsLikeF  
0    0.614525    0.736842  
1    0.614525    0.452632  
2    0.351955    0.400000  
3    0.329609    0.326316  
4    0.351955    0.400000  

[5 rows x 31 columns]

In [37]:
#Adding the column significance to the row significance

for column in numeric_columns:
  df_min_max_scaled[column] = df_min_max_scaled[column] * column_significance[column]

In [38]:
df_min_max_scaled.head(20)

loc_id        date  time isdaytime     tempC     tempF  windspeedMiles  \
0      NaN  2019-06-01  12AM        no  0.012625  0.024177        0.011351   
1      NaN  2019-06-01   1AM        no  0.012625  0.016891        0.011351   
2      NaN  2019-06-01   2AM        no  0.008987  0.016229        0.009241   
3      NaN  2019-06-01   3AM        no  0.008987  0.015235        0.009241   
4      NaN  2019-06-01   4AM        no  0.008987  0.016229        0.007931   
5      NaN  2019-06-01   5AM        no  0.012625  0.024177        0.007931   
6      NaN  2019-06-01   6AM        no  0.019152  0.030139        0.007931   
7      NaN  2019-06-01   7AM        no  0.015407  0.025833        0.007931   
8      NaN  2019-06-01   8AM       yes  0.017547  0.031464        0.007931   
9      NaN  2019-06-01   9AM       yes  0.010057  0.017222        0.007931   
10     NaN  2019-06-01  10AM       yes  0.009415  0.016891        0.009241   
11     NaN  2019-06-01  11AM       yes  0.006527  0.009605        0.011351   
12     NaN  2019-06-01  12PM       yes  0.004066  0.005630        0.013680   
13     NaN  2019-06-01   1PM       yes  0.004066  0.005630        0.011934   
14     NaN  2019-06-01   2PM       yes  0.002140  0.005630        0.010042   
15     NaN  2019-06-01   3PM       yes  0.002140  0.005630        0.010042   
16     NaN  2019-06-01   4PM       yes  0.004066  0.005299        0.010042   
17     NaN  2019-06-01   5PM        no  0.004066  0.005299        0.010042   
18     NaN  2019-06-01   6PM        no  0.006527  0.011923        0.010042   
19     NaN  2019-06-01   7PM        no  0.006313  0.010267        0.011934   

    windspeedKmph  winddirdegree winddir16point  ...  HeatIndexC HeatIndexF  \
0        0.012638       0.144531             SW  ...    0.011439   0.022365   
1        0.012919       0.077083             SW  ...    0.011439   0.014706   
2        0.012919       0.086719             SW  ...    0.007690   0.015012   
3        0.010672       0.096354             SW  ...    0.007690   0.013480   
4        0.008425       0.125260             SW  ...    0.007690   0.015012   
5        0.008987       0.077083             SW  ...    0.011439   0.022365   
6        0.008987       0.134896             SW  ...    0.017784   0.029412   
7        0.008987       0.221614             SW  ...    0.013650   0.023897   
8        0.008987       0.144531            SSW  ...    0.015381   0.027574   
9        0.008425       0.067448            SSW  ...    0.006729   0.012255   
10       0.012919       0.105989            SSW  ...    0.003461   0.008272   
11       0.012638       0.077083            SSW  ...    0.004326   0.005821   
12       0.021204       0.057812            SSW  ...    0.014419   0.013787   
13       0.012217       0.096354            SSW  ...    0.004806   0.013787   
14       0.011796       0.077083            SSW  ...    0.004806   0.013787   
15       0.011374       0.057812            SSW  ...    0.004806   0.007047   
16       0.011796       0.105989            SSW  ...    0.004806   0.007047   
17       0.011796       0.028906            SSW  ...    0.004806   0.013787   
18       0.011796       0.067448            SSW  ...    0.004806   0.013787   
19       0.012217       0.028906              S  ...    0.014419   0.023591   

   DewPointC  DewPointF  WindChillC  WindChillF  WindGustMiles  WindGustKmph  \
0   0.009037   0.012466    0.011769    0.024696       0.012177      0.021144   
1   0.009037   0.016621    0.011769    0.015170       0.012177      0.016327   
2   0.009622   0.016621    0.006741    0.013406       0.013079      0.019538   
3   0.009622   0.015236    0.006313    0.010937       0.013079      0.015256   
4   0.009622   0.015236    0.006741    0.013406       0.008456      0.014453   
5   0.009622   0.015236    0.011769    0.015170       0.008456      0.009368   
6   0.009622   0.016621    0.019152    0.031752       0.007103      0.009903   
7   0.009622   0.014774    0.015300    0.027165       0.007103   

In [39]:
df_num = df_min_max_scaled[numeric_columns]
df_num.drop(["loc_id"], axis=1)

tempC     tempF  windspeedMiles  windspeedKmph  winddirdegree  \
0     0.012625  0.024177        0.011351       0.012638       0.144531   
1     0.012625  0.016891        0.011351       0.012919       0.077083   
2     0.008987  0.016229        0.009241       0.012919       0.086719   
3     0.008987  0.015235        0.009241       0.010672       0.096354   
4     0.008987  0.016229        0.007931       0.008425       0.125260   
...        ...       ...             ...            ...            ...   
1459  0.014016  0.022190        0.007713       0.008285       0.028906   
1460  0.014016  0.031464        0.007713       0.011093       0.105989   
1461  0.017547  0.029145        0.010042       0.011374       0.048177   
1462  0.015407  0.029808        0.010042       0.011374       0.105989   
1463  0.015407  0.025833        0.011934       0.011796       0.028906   

      weatherCode  precipMM  precipInches  humidity  visibilityKm  ...  \
0        0.008892  0.022572      0.002052  0.027781      0.011628  ...   
1        0.008892  0.022572      0.002052  0.023572      0.011628  ...   
2        0.008892  0.022572      0.002052  0.021888      0.011628  ...   
3        0.008892  0.022572      0.002052  0.021888      0.011628  ...   
4        0.008892  0.022572      0.002052  0.027781      0.011628  ...   
...           ...       ...           ...       ...           ...  ...   
1459     0.002122  0.003306      0.002052  0.016837      0.000849  ...   
1460     0.008892  0.022572      0.002052  0.026939      0.011628  ...   
1461     0.008892  0.022572      0.002052  0.023572      0.011628  ...   
1462     0.008892  0.022572      0.002052  0.024413      0.011628  ...   
1463     0.008892  0.022572      0.002052  0.025255      0.011628  ...   

      HeatIndexC  HeatIndexF  DewPointC  DewPointF  WindChillC  WindChillF  \
0       0.011439    0.022365   0.009037   0.012466    0.011769    0.024696   
1       0.011439    0.014706   0.009037   0.016621    0.011769    0.015170   
2       0.007690    0.015012   0.009622   0.016621    0.006741    0.013406   
3       0.007690    0.013480   0.009622   0.015236    0.006313    0.010937   
4       0.007690    0.015012   0.009622   0.015236    0.006741    0.013406   
...          ...         ...        ...        ...         ...         ...   
1459    0.009901    0.015319   0.007442   0.013543    0.014016    0.023637   
1460    0.009901    0.027574   0.007442   0.009696    0.014016    0.033516   
1461    0.015381    0.026654   0.010260   0.016929    0.017547    0.031046   
1462    0.013650    0.026961   0.010260   0.016929    0.015300    0.031752   
1463    0.013650    0.023897   0.009622   0.015236    0.015300    0.027165   

      WindGustMiles  WindGustKmph  FeelsLikeC  FeelsLikeF  
0          0.012177      0.021144    0.012190    0.024192  
1          0.012177      0.016327    0.012190    0.014861  
2          0.013079      0.019538    0.006981    0.013133  
3          0.013079      0.015256    0.006538    0.010714  
4          0.008456      0.014453    0.006981    0.013133  
...             ...           ...         ...         ...  
1459       0.007441      0.011241    0.014517    0.023155  
1460       0.007441      0.014721    0.014517    0.032832  
1461       0.010485      0.014721    0.018174    0.030413  
1462       0.012853      0.013650    0.015847    0.031104  
1463       0.010936      0.015791    0.015847    0.026611  

[1464 rows x 24 columns]

In [40]:
#Creates a list of all non numeric colums
non_numeric = set(df.columns).difference(set(numeric_columns))

In [41]:
df_non_numeric = df[list(non_numeric)]

In [42]:
df_non_numeric.head()

weatherIconUrl winddir16point  time  \
0  https://cdn.worldweatheronline.net/images/wsym...             SW  12AM   
1  https://cdn.worldweatheronline.net/images/wsym...             SW   1AM   
2  https://cdn.worldweatheronline.net/images/wsym...             SW   2AM   
3  https://cdn.worldweatheronline.net/images/wsym...             SW   3AM   
4  https://cdn.worldweatheronline.net/images/wsym...             SW   4AM   

  isdaytime        date    weatherDesc  
0        no  2019-06-01  Partly cloudy  
1        no  2019-06-01  Partly cloudy  
2        no  2019-06-01  Partly cloudy  
3        no  2019-06-01  Partly cloudy  
4        no  2019-06-01  Partly cloudy

In [43]:
df_num.head()

loc_id     tempC     tempF  windspeedMiles  windspeedKmph  winddirdegree  \
0     NaN  0.012625  0.024177        0.011351       0.012638       0.144531   
1     NaN  0.012625  0.016891        0.011351       0.012919       0.077083   
2     NaN  0.008987  0.016229        0.009241       0.012919       0.086719   
3     NaN  0.008987  0.015235        0.009241       0.010672       0.096354   
4     NaN  0.008987  0.016229        0.007931       0.008425       0.125260   

   weatherCode  precipMM  precipInches  humidity  ...  HeatIndexC  HeatIndexF  \
0     0.008892  0.022572      0.002052  0.027781  ...    0.011439    0.022365   
1     0.008892  0.022572      0.002052  0.023572  ...    0.011439    0.014706   
2     0.008892  0.022572      0.002052  0.021888  ...    0.007690    0.015012   
3     0.008892  0.022572      0.002052  0.021888  ...    0.007690    0.013480   
4     0.008892  0.022572      0.002052  0.027781  ...    0.007690    0.015012   

   DewPointC  DewPointF  WindChillC  WindChillF  WindGustMiles  WindGustKmph  \
0   0.009037   0.012466    0.011769    0.024696       0.012177      0.021144   
1   0.009037   0.016621    0.011769    0.015170       0.012177      0.016327   
2   0.009622   0.016621    0.006741    0.013406       0.013079      0.019538   
3   0.009622   0.015236    0.006313    0.010937       0.013079      0.015256   
4   0.009622   0.015236    0.006741    0.013406       0.008456      0.014453   

   FeelsLikeC  FeelsLikeF  
0    0.012190    0.024192  
1    0.012190    0.014861  
2    0.006981    0.013133  
3    0.006538    0.010714  
4    0.006981    0.013133  

[5 rows x 25 columns]

A major problem we face with our data is the fact that most of our data is numeric and not text. All the systems that summarization system used, are specifically made for text data. This means that we have somehow have to make our numeric data behave like text so that we can summarize it. Upto this point, we have been working to make our numeric data behave like text data. And now we have finally reached the last part of the process. From the above output, you can see that there are 61  30x30 matrices. Each 30x30 matrix is a result of us calculating the signnificance of our numeric columns against themselves. 1464 is the number of rows. These 30X30 matrices are important since they will help us get the embeddings for our numeric data as we pass them through an RNN model.

In [44]:
'''Preprocessing the text part of our data'''

'Preprocessing the text part of our data'

In [45]:
#Non numeric data
df_non_numeric.head()

weatherIconUrl winddir16point  time  \
0  https://cdn.worldweatheronline.net/images/wsym...             SW  12AM   
1  https://cdn.worldweatheronline.net/images/wsym...             SW   1AM   
2  https://cdn.worldweatheronline.net/images/wsym...             SW   2AM   
3  https://cdn.worldweatheronline.net/images/wsym...             SW   3AM   
4  https://cdn.worldweatheronline.net/images/wsym...             SW   4AM   

  isdaytime        date    weatherDesc  
0        no  2019-06-01  Partly cloudy  
1        no  2019-06-01  Partly cloudy  
2        no  2019-06-01  Partly cloudy  
3        no  2019-06-01  Partly cloudy  
4        no  2019-06-01  Partly cloudy

In [46]:
df_non_numeric = df_non_numeric.drop(["weatherIconUrl"], axis=1)
df_non_numeric
#We droped the icon url since we do not need it for out use case

winddir16point  time isdaytime        date           weatherDesc
0                SW  12AM        no  2019-06-01         Partly cloudy
1                SW   1AM        no  2019-06-01         Partly cloudy
2                SW   2AM        no  2019-06-01         Partly cloudy
3                SW   3AM        no  2019-06-01         Partly cloudy
4                SW   4AM        no  2019-06-01         Partly cloudy
...             ...   ...       ...         ...                   ...
1459              W   7PM        no  2019-07-31  Patchy rain possible
1460            WSW   8PM        no  2019-07-31         Partly cloudy
1461            WSW   9PM        no  2019-07-31         Partly cloudy
1462            WSW  10PM        no  2019-07-31         Partly cloudy
1463            WSW  11PM        no  2019-07-31         Partly cloudy

[1464 rows x 5 columns]

In [47]:
#Get a list of words
all_words = []
for column in df_non_numeric.columns:
  column_words = set(df_non_numeric[column])
  all_words.extend(list(column_words))

print(all_words)

['NW', 'NNE', 'S', 'SE', 'WSW', 'NNW', 'N', 'WNW', 'W', 'NE', 'SSW', 'ENE', 'SW', 'E', 'SSE', 'ESE', '6AM', '7PM', '5AM', '3AM', '1PM', '8AM', '1AM', '4PM', '2AM', '9AM', '12PM', '9PM', '3PM', '2PM', '5PM', '4AM', '6PM', '11AM', '11PM', '12AM', '10AM', '8PM', '7AM', '10PM', 'yes', 'no', '2019-06-16', '2019-07-31', '2019-06-20', '2019-07-23', '2019-06-23', '2019-06-22', '2019-06-01', '2019-07-10', '2019-06-02', '2019-07-11', '2019-07-27', '2019-07-06', '2019-06-08', '2019-07-08', '2019-07-07', '2019-06-27', '2019-06-15', '2019-07-25', '2019-07-24', '2019-07-17', '2019-07-15', '2019-06-07', '2019-06-03', '2019-07-21', '2019-07-22', '2019-07-01', '2019-06-30', '2019-07-02', '2019-07-12', '2019-07-04', '2019-06-25', '2019-06-09', '2019-06-13', '2019-06-12', '2019-06-21', '2019-06-26', '2019-07-09', '2019-07-16', '2019-06-10', '2019-06-18', '2019-06-24', '2019-07-18', '2019-06-19', '2019-07-30', '2019-06-05', '2019-07-28', '2019-06-28', '2019-07-13', '2019-06-04', '2019-06-06', '2019-07-03'

In [48]:
#Give each entry a number that will represent it
WORD_2_TOKEN = {}
TOKEN_2_WORD = {}

def Tokenize(word, i):
  WORD_2_TOKEN[word] = i
  TOKEN_2_WORD[i] = word
  return

def encode(row):
  encoded = []
  for word in row:
    encoded.append(WORD_2_TOKEN[word])
  return encoded

def decode(row_encoded):
  decoded = []
  for word in row_encoded:
    decoded.append(TOKEN_2_WORD[word])
  return decoded


In [49]:
for i in range(len(all_words)):
  Tokenize(all_words[i], i)

In [50]:
print(f"Encode example: {encode(['Light rain shower','Partly cloudy'])}")

Encode example: [106, 109]


In [51]:
print(f"Decode example: {decode([1,6])}")

Decode example: ['NNE', 'N']


Now our text data is ready to be passed to a deep learning model.

During training, we will pass the numerical representation for our word through an embedding layer which will produce word embeddings. As for the part concerning numeric data, we will take our 26X26 matrix and put it through a softmax layer and then through a linear layer to obtain the embeddings.
